src/sys/wait.rs

```diff
@@ -258,34 +261,38 @@ mod status {
     }
 }
 
-fn decode(pid : Pid, status: i32) -> WaitStatus {
-    if status::exited(status) {
-        WaitStatus::Exited(pid, status::exit_status(status))
-    } else if status::signaled(status) {
-        WaitStatus::Signaled(pid, status::term_signal(status), status::dumped_core(status))
-    } else if status::stopped(status) {
-        cfg_if! {
-            if #[cfg(any(target_os = "linux", target_os = "android"))] {
-                fn decode_stopped(pid: Pid, status: i32) -> WaitStatus {
-                    let status_additional = status::stop_additional(status);
-                    if status::syscall_stop(status) {
-                        WaitStatus::PtraceSyscall(pid)
-                    } else if status_additional == 0 {
-                        WaitStatus::Stopped(pid, status::stop_signal(status))
-                    } else {
-                        WaitStatus::PtraceEvent(pid, status::stop_signal(status), status::stop_additional(status))
+impl WaitStatus {
+    pub fn from_raw(pid : Pid, status: i32) -> Result<WaitStatus> {
+        Ok(if status::exited(status) {
+            WaitStatus::Exited(pid, status::exit_status(status))
+        } else if status::signaled(status) {
+            WaitStatus::Signaled(pid, try!(status::term_signal(status)), status::dumped_core(status))
+        } else if status::stopped(status) {
+            cfg_if! {
```

Comment:
This would read better if we can get rid of this helper function. If not, let's move this helper function out of this function and instead into the module (and add good docs for it even though it's private).